runtime/memprof.c

```diff
@@ -16,23 +16,531 @@
+#define One_log1m_lambda(config) \
+  Double_val(Field(config, CONFIG_FIELD_1LOG1ML))
+
+/* The number of frames to record for each allocation site */
+#define Callstack_size(config) \
+  Int_val(Field(config, CONFIG_FIELD_FRAMES))
+
+/* callbacks */
+#define Alloc_minor(config)   Field(config, CONFIG_FIELD_ALLOC_MINOR)
+#define Alloc_major(config)   Field(config, CONFIG_FIELD_ALLOC_MAJOR)
+#define Promote(config)       Field(config, CONFIG_FIELD_PROMOTE)
+#define Dealloc_minor(config) Field(config, CONFIG_FIELD_DEALLOC_MINOR)
+#define Dealloc_major(config) Field(config, CONFIG_FIELD_DEALLOC_MAJOR)
+
+/* Callback indexes. "Major" and "minor" are not distinguished here. */
+
+#define CB_NONE          0
+#define CB_ALLOC         1
+#define CB_PROMOTE       2
+#define CB_DEALLOC       3
+
+/* Maximum value of a callback index */
+#define CB_MAX           CB_DEALLOC
+
+/* How many bits required for a callback index */
+#define CB_BITS          2
+
+/* the mask for a given callback index */
+#define CB_MASK(cb) (1 << ((cb) - 1))
+
+/* Structure for each tracked allocation. Six words (with many spare
+ * bits in the final word). */
+
+struct entry_s {
+  /* Memory block being sampled. This is a weak GC root. Note that
+   * during the allocation callback of a block allocated directly by OCaml,
+   * this may be a comballoc offset. */
+  value block;
+
+  /* The value returned by the previous callback for this block, or
+   * the callstack (as a value-tagged pointer to the C heap) if the
+   * alloc callback has not been called yet.  This is a strong GC
+   * root. */
+  value user_data;
+
+  /* Number of samples in this block. */
+  size_t samples;
+
+  /* The size of this block. */
+  size_t wosize;
+
+  /* The thread currently running a callback for this entry,
+   * or NULL if there is none */
+  memprof_thread_t running;
+
+  /* Is `block` actually an offset? */
+  unsigned int offset : 1;
+
+  /* Was this block initially allocated in the minor heap? */
+  unsigned int alloc_young : 1;
+
+  /* The source of the allocation: normal allocations or custom_mem (SRC_*). */
+  unsigned int source : 2;
+
+  /* Has this block been promoted? Implies [alloc_young]. */
+  unsigned int promoted : 1;
+
+  /* Has this block been deallocated? */
+  unsigned int deallocated : 1;
+
+  /* Which callback (CB_*) is currently running for this entry. */
+  unsigned int callback : CB_BITS;
+
+  /* A mask of callbacks (1 << (CB_* - 1)) which have been called (not
+   * necessarily completed) for this entry. */
+  unsigned int callbacks : CB_MAX;
+
+  /* Has this entry been marked for deletion. */
+  unsigned int deleted : 1;
+
+  /* There are a number of spare bits here for future expansion,
+   * without increasing the size of an entry */
+};
+
+/* A resizable array of entry_s entries. */
+
+struct entries_s {
+  entry_t t; /* Pointer to array of entry_s structures */
+  size_t min_size, size, live; /* array allocation management */
```

Comment:
```suggestion
  size_t min_capacity, capacity, size; /* array allocation management */
```

I'm suggesting this, because "live" can be misunderstood with an hypothetical index of live entries in the sense of GC (just like we have an index for young entries).

Also, size/capacity is somewhat standard terminology used for C++ vector. I find it somewhat confusing to use `size` something which is not what C++ vector calls `size`.

Of course, this is just a suggestion.

Reply:
I have no objection to this change (a slight dislike for your `size` but I don't like my `live` either), so will make it.

Reply:
`len` instead of `size`?